src/qiboml/models/decoding.py

```diff
@@ -283,3 +286,38 @@ def output_shape(self) -> tuple[int, int]:
     @property
     def analytic(self) -> bool:  # pragma: no cover
         return False
+    
+@dataclass
+class VQLS(QuantumDecoding):
+    """Decoder for the Variational Quantum Linear Solver (VQLS).
+
+    Args:
+        target_state (ndarray): Target solution vector |b⟩.
+        A (ndarray): The matrix A in the linear system A|x⟩ = |b⟩.
+    """
+    target_state: ndarray = None
+    A: ndarray = None
+
+    def __post_init__(self):
+        super().__post_init__()
+        self.target_state = torch.tensor(self.target_state, dtype=torch.complex128)
+        self.A = torch.tensor(self.A, dtype=torch.complex128)
+
+    def __call__(self, circuit: Circuit) -> torch.Tensor:
+        result = super().__call__(circuit)
+        state = result.state()  
+        final_state = self.A @ state
+        normalized = final_state / torch.linalg.norm(final_state)
+        overlap = torch.vdot(self.target_state, normalized)
+        return 1 - overlap.abs() ** 2
+
```

Comment:
Please double check the suggestion here. In particular, I am not sure whether we should use here the `scipy.linalg` module instead of the `self.backend.np.linalg` one.

```suggestion
        super().__post_init__()
        self.target_state = self.backend.cast(self.target_state, dtype=self.backend.np.complex128)
        self.A = self.cast(self.A, dtype=self.backend.np.complex128)

    def __call__(self, circuit: Circuit) -> torch.Tensor:
        result = super().__call__(circuit)
        state = result.state()  
        final_state = self.A @ state
        normalized = final_state / self.backend.np.linalg.norm(final_state)
        overlap = self.backend.np.vdot(self.target_state, normalized)
        return 1 - overlap.abs() ** 2

```

Reply:
To add, `backend.np.linalg.norm` can be replaced with the simpler `backend.calculate_vector_norm` (or `backend.calculate_matrix_norm` when required).